### ISSUE TITLE
Only partial downloads when youtube-dl extension enabled

### ISSUE THREAD

**Issue by computerfreak105** [gpodder/gpodder#917]
GPO 3.10.17 under Windows 10 pro, 64-bit. I'd like to download youtube channels using the extension only for downloading. I get partial files with different sizes but the process doesn't complete. It should download audio only m4a and rename it with the date at the begin of file names. Here a log file and my settings.json as pdf. Sadly I don't have any programming talents. So thanks for looking into this.
[2020-12-08 file not found.log](https://github.com/gpodder/gpodder/files/5661032/2020-12-08.file.not.found.log)
[Settings json.pdf](https://github.com/gpodder/gpodder/files/5661044/Settings.json.pdf)

**Comment by auouymous** [comment]
You can't download audio-only formats from youtube yet, because gpodder expects youtube channels to always be video. I have patches that will be revisited and posted after the next release, which should be in the next couple days or so.

**Comment by computerfreak105** [comment]
Any chance these patches will be posted soon? Would make for a nice Christmas present :-)
Thank you very much and merry Christmas
Ben

**Comment by elelay** [comment]
Selecting for instance "worstaudio" as the youtube ids seems to download a .webm containing opus audio only.

**Comment by auouymous** [comment]
Yes, it is possible to manually select an audio format, and my one patch makes that easier by adding the formats to the selector. The biggest problem was that gpodder only wants to use the video player (and video icon) for youtube episodes. Anyone using a video player that lacks a UI will find it difficult to control audio-only youtube episodes. You could of course just change the video player to an audio player if you only want audio for everything. I haven't done this yet, but an extension or builtin to allow per-channel youtube formats would allow some to be audio and others to be video, or even different video resolutions per channel.

**Comment by elelay** [comment]
Ha ok: your patches are welcome, then :-)

**Comment by tpikonen** [comment]
I'm working on making the youtube-dl extension support sites other than youtube, and I also noticed the lack of audio formats.

There is a comment in youtube-dl.py / add_format() telling that format strings like 'bestaudio' work in youtube.preferred_fmt_ids. It  does work when downloading with the extension (but only when youtube.preferred_fmt_id is set to '0'). But, streaming from youtube channels uses code from youtube.py, and it breaks with a non-integer format.

I already implemented a separate config var for youtube-dl format string before I noticed this issue, but maybe @auouymous has a better solution?

**Comment by auouymous** [comment]
@tpikonen Every site has its own set of IDs, and would need separate {site}.preferred_fmt_id/ids config options.

preferred_fmt_ids is only ever valid when preferred_fmt_id is zero. preferred_fmt_id is selected from the drop-down menu and chooses a range of IDs in youtube.py's format_dict, while setting it to zero unlocks custom user-defined IDs in preferred_fmt_ids. Youtube streaming (HLS) does not use youtube-dl or the preferred_fmt_id/ids variables. It uses preferred_**hls**_fmt_id/ids and no comment says you can use a best string for it. HLS formats are defined in youtube.py's hls_formats_dict.

The patches at http://borg.qzx.com/files/gpodder/ add best formats for youtube-dl and audio formats for youtube or youtube-dl. I think the best-formats patch was waiting for #796 to be merged, and I could probably open a PR now, but I need to verify it works properly. The audio-formats patch works but has a few issues I haven't fixed yet. #872 can cause incorrect icon to be used for audio files, although a quick test shows it to now be working correctly for me. The biggest problem is that gpodder uses a weba ext (audio/webm) while youtube-dl uses webm ext (video/webm), which causes the video icon to show and video player to be used. I think the unfinished mime-type patch was going to fix that and force video-only streaming.

The per-channel formats could be done with a simple audio/video checkbox for each youtube channel and separate audio/video format drop-downs in the video preferences. Or separate preferred_fmt_id/ids, format drop-down and custom format text input on each channel dialog, and use the global setting when channel format is id=0 and ids=''. The latter is more complex but allows different formats per channel and it might be possible to reuse this for non-youtube youtube-dl sites, but I imagine any site that could use youtube-dl would need custom support to download a list of episodes.

**Comment by tpikonen** [comment]
I'm not sure I've understood the problem, but in general the video format lists (3 of them!) are already rather long. IMO it would be better to just have a resolution dropdown for video and possibly a bandwidth dropdown for audio and deduce the actual formats from those.

The video/audio checkbox per channel would perhaps be useful. In #949 I implemented a config var which forces streaming using the video player, mostly because there is no way to tell if a youtube-dl URL points to video or audio. Then again, there are youtube channels with both video and audio content, so launching different players by a channel-based setting does not work for those.

**Comment by auouymous** [comment]
Resolution and bandwidth doesn't work because some users can't play the webm formats. Selecting the MP4 or M4A format only downloads those types, while WEBM/WEBA formats will download either type. But audio-only formats shouldn't be allowed until youtube-dl can properly identify them as audio.

Youtube is a video platform, the audio-only and video-only formats are only there for adaptive players (https://en.wikipedia.org/wiki/Dynamic_Adaptive_Streaming_over_HTTP) to scale playback. The audio-only formats even exist for videos that were uploaded without audio. This is another issue with supporting audio-only formats because gpodder tries to download them like any other file. Try downloading the same episode with both gpodder and the youtube-dl extension, gpodder is so much slower because it doesn't properly support DASH. The same also applies with the video-only formats.
